src/main/kotlin/uk/gov/dluhc/printapi/database/TwelveByteId.kt

```diff
@@ -0,0 +1,157 @@
+        timestamp,
+        RANDOM_VALUE1,
+        RANDOM_VALUE2,
+        counter,
+        checkCounter
+    )
+
+    private constructor(timestamp: Int, randomValue1: Int, randomValue2: Short, counter: Int, checkCounter: Boolean) {
+        require(randomValue1 and -0x1000000 == 0) { "The random value must be between 0 and 16777215 (it must fit in three bytes)." }
+        require(!(checkCounter && counter and -0x1000000 != 0)) { "The counter must be between 0 and 16777215 (it must fit in three bytes)." }
+        this.timestamp = timestamp
+        this.counter = counter and LOW_ORDER_THREE_BYTES
+        this.randomValue1 = randomValue1
+        this.randomValue2 = randomValue2
+    }
+
+    override fun toString(): String = toHexString()
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as TwelveByteId
+
+        if (timestamp != other.timestamp) return false
+        if (counter != other.counter) return false
+        if (randomValue1 != other.randomValue1) return false
+        if (randomValue2 != other.randomValue2) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = timestamp
+        result = 31 * result + counter
+        result = 31 * result + randomValue1
+        result = 31 * result + randomValue2
+        return result
+    }
+
+    private fun toHexString(): String {
+        val chars = CharArray(OBJECT_ID_LENGTH * 2)
+        var i = 0
+        for (b in toByteArray()) {
+            chars[i++] = HEX_CHARS[b.toInt() shr 4 and 0xF]
+            chars[i++] = HEX_CHARS[b.toInt() and 0xF]
+        }
+        return String(chars)
+    }
+
+    private fun toByteArray(): ByteArray {
+        val buffer = ByteBuffer.allocate(OBJECT_ID_LENGTH)
+        putToByteBuffer(buffer)
+        return buffer.array()
+    }
+
+    private fun putToByteBuffer(buffer: ByteBuffer) {
+        requireNotNull(buffer) { "buffer cannot be null" }
+        require(buffer.remaining() >= OBJECT_ID_LENGTH) { "buffer.remaining() >=12" }
+
+        buffer
+            .put(int3(timestamp))
+            .put(int2(timestamp))
+            .put(int1(timestamp))
+            .put(int0(timestamp))
+            .put(int2(randomValue1))
+            .put(int1(randomValue1))
+            .put(int0(randomValue1))
+            .put(short1(randomValue2))
+            .put(short0(randomValue2))
+            .put(int2(counter))
+            .put(int1(counter))
+            .put(int0(counter))
+    }
+
+    private fun int3(x: Int): Byte = (x shr 24).toByte()
+
+    private fun int2(x: Int): Byte = (x shr 16).toByte()
+
+    private fun int1(x: Int): Byte = (x shr 8).toByte()
+
+    private fun int0(x: Int): Byte = x.toByte()
+
+    private fun short1(x: Short): Byte = (x.toInt() shr 8).toByte()
+
+    private fun short0(x: Short): Byte = x.toByte()
```

Comment:
Cannot comment on the correctness of the algorithm - it would be easy to have a typo somewhere.
Although I understand point about unit testing this class is dififcult due to all the randomness - as a one of test say 10,000 random values could be generated and verified no clashes occured.  (I'm sure you must have done some verification already)